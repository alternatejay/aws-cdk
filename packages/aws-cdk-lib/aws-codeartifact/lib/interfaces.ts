import * as events from '@aws-cdk/aws-events';
import * as iam from '@aws-cdk/aws-iam';
import * as kms from '@aws-cdk/aws-kms';
import { IResource } from '@aws-cdk/core';

/**
 * Represents a CodeArtifact repository
 * @experimental
 */
export interface IRepository extends IResource {

  /**
     * The ARN of repository resource.
     * Equivalent to doing `{ 'Fn::GetAtt': ['LogicalId', 'Arn' ]}`
     * in CloudFormation if the underlying CloudFormation resource
     * surfaces the ARN as a return value -
     * if not, we usually construct the ARN "by hand" in the construct,
     * using the Fn::Join function.
     *
     * It needs to be annotated with '@attribute' if the underlying CloudFormation resource
     * surfaces the ARN as a return value.
     *
     * @default Empty string
     * @attribute
     */
  readonly repositoryArn?: string;
  /**
       * The physical name of the repository resource.
       * Often, equivalent to doing `{ 'Ref': 'LogicalId' }`
       * (but not always - depends on the particular resource modeled)
       * in CloudFormation.
       * Also needs to be annotated with '@attribute'.
       * @default Empty string
       * @attribute
       */
  readonly repositoryName: string;
  /**
     * Name the repository from CFN attribute
       * @default Empty string
       * @attribute
     */
  readonly repositoryNameAttr?: string,
  /**
      * A text description of the repository.
      * @attribute
      */
  readonly repositoryDescription?: string;
  /**
       * The domain repository owner
       * Often, equivalent to the account id.
       * @default Empty string
       * @attribute
       */
  readonly repositoryDomainOwner?: string;
  /**
       * The domain the repository belongs to
       * @default Empty string
       * @attribute
       */
  readonly repositoryDomainName?: string;
  /**
       * Grants the given IAM identity permissions to read from the repository
       */
  grantRead(identity: iam.IGrantable): iam.Grant;
  /**
       * Grants the given IAM identity permissions to write from the repository
       */
  grantWrite(identity: iam.IGrantable): iam.Grant;
  /**
       * Grants the given IAM identity permissions to read/write from the repository
       */
  grantReadWrite(identity: iam.IGrantable): iam.Grant;
  /**
       * Defines a CloudWatch event rule which triggers for repository events. Use
       * `rule.addEventPattern(pattern)` to specify a filter.
       */
  onEvent(id: string, options?: events.OnEventOptions): events.Rule;
  /**
       * Defines a CloudWatch event rule which triggers when a "CodeArtifact Package
       *  Version State Change" event occurs.
       */
  onPackageVersionStateChange(id: string, options?: events.OnEventOptions): events.Rule;

  /**
   * Set the domain for the repository
   * @param domain The domain the repository is apart of
   */
  assignDomain(domain: IDomain): void;
}

/**
 * Reference to a repository
 * @experimental
 */
export interface RepositoryAttributes {

  /**
     * The ARN of repository resource.
     * Equivalent to doing `{ 'Fn::GetAtt': ['LogicalId', 'Arn' ]}`
     * in CloudFormation if the underlying CloudFormation resource
     * surfaces the ARN as a return value -
     * if not, we usually construct the ARN "by hand" in the construct,
     * using the Fn::Join function.
     *
     * It needs to be annotated with '@attribute' if the underlying CloudFormation resource
     * surfaces the ARN as a return value.
     * @default Empty string
     * @attribute
     */
  readonly repositoryArn?: string;
  /**
       * The physical name of the repository resource.
       * Often, equivalent to doing `{ 'Ref': 'LogicalId' }`
       * (but not always - depends on the particular resource modeled)
       * in CloudFormation.
       * Also needs to be annotated with '@attribute'.
       * @default Empty string
       * @attribute
       */
  readonly repositoryName?: string;
  /**
   * A text description of the repository.
   * @default Empty string
   */
  readonly repositoryDescription?: string;
  /**
       * The domain repository owner
       * Often, equivalent to the account id.
       * @default Empty string
       * @attribute
       */
  readonly repositoryDomainOwner?: string;
  /**
       * The domain the repository belongs to
       * @default Empty string
       * @attribute
       */
  readonly repositoryDomainName?: string;
}

/**
 * Represents a CodeArtifact domain
 * @experimental
 */
export interface IDomain extends IResource {
  /**
* The ARN of domain resource.
* @attribute
*/
  readonly domainArn: string;

  /**
   * The physical name of the domain resource.
   * @attribute
   */
  readonly domainName?: string;
  /**
   * The attribute name of the domain resource.
   * @attribute
   */
  readonly domainNameAttr?: string;

  /**
   * The domain owner
   * Often, equivalent to the account id.
   * @attribute
   */
  readonly domainOwner?: string;

  /**
   * The KMS encryption key used for the domain resource.
   * @default AWS Managed Key
   * @attribute
   */
  readonly domainEncryptionKey?: kms.IKey;

  /**
   * Resource policy for the domain
   */
  readonly policyDocument?: iam.PolicyDocument
}

/**
 * Reference to a domain
 * @experimental
 */
export interface DomainAttributes {
  /**
 * The ARN of domain resource.
 * Equivalent to doing `{ 'Fn::GetAtt': ['LogicalId', 'Arn' ]}`
 * in CloudFormation if the underlying CloudFormation resource
 * surfaces the ARN as a return value -
 * if not, we usually construct the ARN "by hand" in the construct,
 * using the Fn::Join function.
 *
 * It needs to be annotated with '@attribute' if the underlying CloudFormation resource
 * surfaces the ARN as a return value.
 *
 * @attribute
 */
  readonly domainArn: string;

  /**
   * The physical name of the domain resource.
   * Often, equivalent to doing `{ 'Ref': 'LogicalId' }`
   * (but not always - depends on the particular resource modeled)
   * in CloudFormation.
   * Also needs to be annotated with '@attribute'.
   * @default Empty string
   *
   * @attribute
   */
  readonly domainName?: string;

  /**
   * The domain owner
   * Often, equivalent to the account id.
   * @default Empty string
   * @attribute
   */
  readonly domainOwner?: string;

  /**
   * The KMS encryption key used for the domain resource.
   * @default AWS Managed Key
   * @attribute
   */
  readonly domainEncryptionKey?: kms.IKey;
}