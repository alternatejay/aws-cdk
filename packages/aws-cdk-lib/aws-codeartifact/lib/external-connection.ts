/**
 * CodeArtifact supports an external connection to the following public repositories.
 * @experimental
 * @see https://docs.aws.amazon.com/codeartifact/latest/ug/external-connection.html#supported-public-repositories
 */
export enum ExternalConnection {
  /**
   * NPM public registry
   */
  NPM = 'public:npmjs',
  /**
   * NuGet.org public registry
   */
  DOTNET_NUGETORG = 'public:nuget-org',
  /**
   * Python Package Index
   */
  PYTHON_PYPI = 'public:pypi',
  /**
   * Maven Central
   */
  MAVEN_CENTRAL = 'public:maven-central',
  /**
   * Google Android repository
   */
  MAVEN_GOOGLEANDROID = 'public:maven-googleandroid',
  /**
   * Gradle plugins repository
   */
  MAVEN_GRADLEPLUGINS = 'public:maven-gradleplugins',
  /**
   * CommonsWare Android repository
   * */
  MAVEN_COMMONSWARE = 'public:maven-commonsware',
}