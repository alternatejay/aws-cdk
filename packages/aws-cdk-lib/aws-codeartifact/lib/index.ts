// AWS::CodeArtifact CloudFormation Resources:
export * from './codeartifact.generated';
export * from './domain';
export * from './repository';
export * from './external-connection';
export * from './interfaces';