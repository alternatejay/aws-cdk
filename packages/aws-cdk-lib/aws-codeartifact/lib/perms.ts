/**
 * @experimental
 */
export const REPOSITORY_WRITE_ACTIONS = [
  'codeartifact:PublishPackageVersion',
  'codeartifact:PutPackageMetadata',
];

/**
 * @experimental
 */
export const REPOSITORY_READ_ACTIONS = [
  'codeartifact:DescribePackageVersion',
  'codeartifact:DescribeRepository',
  'codeartifact:GetPackageVersionReadme',
  'codeartifact:GetRepositoryEndpoint',
  'codeartifact:ListPackages',
  'codeartifact:ListPackageVersions',
  'codeartifact:ListPackageVersionAssets',
  'codeartifact:ListPackageVersionDependencies',
  'codeartifact:ReadFromRepository',
];

/**
 * @experimental
 */
export const DOMAIN_READ_ACTIONS = [
  'codeartifact:GetDomainPermissionsPolicy',
  'codeartifact:ListRepositoriesInDomain',
  'codeartifact:DescribeDomain',
];

/**
 * @experimental
 */
export const DOMAIN_CREATE_ACTIONS = [
  'codeartifact:CreateRepository',
];

/**
 * @experimental
 */
export const DOMAIN_LOGIN_ACTIONS = [
  'codeartifact:GetAuthorizationToken',
];