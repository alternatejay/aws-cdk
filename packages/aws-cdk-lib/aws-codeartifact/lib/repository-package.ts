// https://docs.aws.amazon.com/codeartifact/latest/APIReference/API_GetRepositoryEndpoint.html
export enum PackageFormat {
    ANY = "*",
    CARGO = "cargo",
    GENERIC = "generic",
    MAVEN = "maven",
    NPM = "npm",
    NUGET = "nuget",
    PYPI = "pypi",
    RUBY = "ruby",
    SWIFT = "swift"
}
