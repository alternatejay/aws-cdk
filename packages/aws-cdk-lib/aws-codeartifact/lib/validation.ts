/**
 * Throws an error if a value is defined and not an integer or not in a range.
 */
export function validate(name: string,
  rule : { documentationLink: string, required?: boolean, pattern? : RegExp, minLength? : number, maxLength?: number },
  value?: string) {

  if (!value) {
    if (rule.required) {
      throw new Error(`'${name}' is required and cannot be empty`);
    }
    return;
  }

  if (rule.pattern && !rule.pattern.test(value)) {
    throw new Error(`${name}: must match pattern ${rule.pattern.source}. Must match rules from ${rule.documentationLink}`);
  }

  if (rule.minLength && value.length < rule.minLength) {
    throw new Error(`${name}: must be at least ${rule.minLength} characters long. Must match rules from ${rule.documentationLink}`);
  }

  if (rule.maxLength && value.length > rule.maxLength) {
    throw new Error(`${name}: must be less than ${rule.maxLength} characters long. Must match rules from ${rule.documentationLink}`);
  }
}
