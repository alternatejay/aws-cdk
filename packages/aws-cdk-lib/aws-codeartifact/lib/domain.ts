import * as iam from '@aws-cdk/aws-iam';
import * as kms from '@aws-cdk/aws-kms';
import { Resource, Stack, Lazy, Token } from '@aws-cdk/core';
import { Construct } from 'constructs';
import { CfnDomain } from './codeartifact.generated';
import { IDomain, IRepository, DomainAttributes } from './interfaces';
import { DOMAIN_CREATE_ACTIONS, DOMAIN_LOGIN_ACTIONS, DOMAIN_READ_ACTIONS } from './perms';
import { validate } from './validation';

/**
 * Properties for a new CodeArtifact domain
 * @experimental
 */
export interface DomainProps {
  /**
  * The name of the domain
  * @default Unique id
  */
  readonly domainName?: string;

  /**
   * The KMS encryption key used for the domain resource.
   * @default AWS Managed Key
   * @attribute
   */
  readonly domainEncryptionKey?: kms.IKey;
  /**
     * Principal for the resource policy for the domain
     * @default AccountRootPrincipal
     */
  readonly principal?: iam.IPrincipal

  /**
   * Resource policy for the domain
   * @default Open policy that allows principal to reader, create, and generate authorization token.
   */
  readonly policyDocument?: iam.PolicyDocument
}

/**
 * A new CodeArtifacft domain
 * @experimental
 */
export class Domain extends Resource implements IDomain {
  /**
 * Import an existing domain provided an ARN
 *
 * @param scope The parent creating construct
 * @param id The construct's name
 * @param domainArn Domain ARN (i.e. arn:aws:codeartifact:us-east-2:444455556666:domain/MyDomain)
 */
  public static fromDomainArn(scope: Construct, id: string, domainArn: string): IDomain {
    return Domain.fromDomainAttributes(scope, id, { domainArn: domainArn });
  }

  /**
   * Import an existing domain
   */
  public static fromDomainAttributes(scope: Construct, id: string, attrs: DomainAttributes): IDomain {
    const stack = Stack.of(scope);
    const domainName = attrs.domainName || stack.parseArn(attrs.domainArn).resourceName;

    if (!domainName || domainName == '') {
      throw new Error('Domain name is required as a resource name with the ARN');
    }

    class Import extends Resource implements IDomain {
      domainArn: string = '';
      domainName?: string | undefined;
      domainOwner?: string | undefined;
      domainEncryptionKey?: kms.IKey | undefined;
    }

    const instance = new Import(scope, id);
    instance.domainName = domainName;
    instance.domainArn = attrs.domainArn;
    instance.domainEncryptionKey = attrs.domainEncryptionKey;

    return instance;
  }

  public readonly domainName?: string;
  public readonly domainNameAttr?: string;
  public readonly domainArn: string = '';
  public readonly domainOwner?: string = '';
  public readonly domainEncryptionKey?: kms.IKey;
  public readonly policyDocument?: iam.PolicyDocument;
  private readonly cfnDomain: CfnDomain;

  constructor(scope: Construct, id: string, props?: DomainProps) {
    super(scope, id);

    // Set domain and encryption key as we will validate them before creation
    const domainName = props?.domainName ?? this.node.uniqueId;
    const domainEncryptionKey = props?.domainEncryptionKey ?? null;

    this.validateProps(domainName, domainEncryptionKey);

    // Create the CFN domain instance
    this.cfnDomain = new CfnDomain(this, 'Resource', {
      domainName: domainName,
      permissionsPolicyDocument: Lazy.anyValue({ produce: () => props?.policyDocument?.toJSON() }),
      encryptionKey: domainEncryptionKey?.keyId,
    });

    this.domainName = domainName;
    this.domainNameAttr = this.cfnDomain.attrName;
    this.domainArn = this.cfnDomain.attrArn;
    this.domainOwner = this.cfnDomain.attrOwner;
    this.policyDocument = props?.policyDocument;
  }

  /**
   * Add a repositories to the domain
   */
  addRepositories(...repositories: IRepository[]): IDomain {
    if (repositories.length > 0) {
      repositories.forEach(r => r.assignDomain(this));
    }

    return this;
  }

  private validateProps(domainName : string, domainEncryptionKey? : kms.IKey | null) {
    if (Token.isUnresolved(domainName)) {
      throw new Error(`'domainName' must resolve, got: '${domainName}'`);
    }

    validate('DomainName',
      { required: true, minLength: 2, maxLength: 50, pattern: /[a-z][a-z0-9\-]{0,48}[a-z0-9]/gi, documentationLink: 'https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codeartifact-domain.html#cfn-codeartifact-domain-domainname' },
      domainName);

    validate('EncryptionKey',
      { minLength: 1, maxLength: 2048, pattern: /\S+/gi, documentationLink: 'https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codeartifact-domain.html#cfn-codeartifact-domain-encryptionkey' },
      domainEncryptionKey?.keyArn || '');
  }

  /**
   * Adds a statement to the IAM resource policy associated with this domain.
   */
  public addToResourcePolicy(statement: iam.PolicyStatement): iam.AddToResourcePolicyResult {

    if (!this.policyDocument) {
      const p = this.policyDocument || new iam.PolicyDocument();

      p.addStatements(statement);

      return { statementAdded: true, policyDependable: p };
    }

    return { statementAdded: false };
  }

  private grant(principal: iam.IGrantable, iamActions: string[], resource: string = '*'): iam.Grant {
    return iam.Grant.addToPrincipalOrResource({
      grantee: principal,
      actions: iamActions,
      resourceArns: [resource],
      resource: this,
    });
  }

  /**
   * Assign default login, creation, and read for the domain.
   * @param principal The principal of for the policy
   * @see https://docs.aws.amazon.com/codeartifact/latest/ug/domain-policies.html#domain-policy-example
   */
  grantDefaultPolicy(principal: iam.IPrincipal): iam.Grant {
    const p = principal;
    this.grantLogin(p);
    this.grantCreate(p);
    return this.grantRead(p);
  }

  /**
     * Adds read actions for the principal to the domain's
     * resource policy
     * @param principal The principal for the policy
     * @see https://docs.aws.amazon.com/codeartifact/latest/ug/domain-spolicies.html
     */
  public grantRead(principal: iam.IPrincipal): iam.Grant {
    return this.grant(principal, DOMAIN_READ_ACTIONS);
  }
  /**
     * Adds GetAuthorizationToken for the principal to the domain's
     * resource policy
     * @param principal The principal for the policy
     * @see https://docs.aws.amazon.com/codeartifact/latest/ug/domain-policies.html
     */
  public grantLogin(principal: iam.IPrincipal): iam.Grant {
    return this.grant(principal, DOMAIN_LOGIN_ACTIONS);
  }
  /**
     * Adds CreateRepository for the principal to the domain's
     * resource policy
     * @param principal The principal for the policy
     * @see https://docs.aws.amazon.com/codeartifact/latest/ug/domain-policies.html
     */
  public grantCreate(principal: iam.IPrincipal): iam.Grant {
    return this.grant(principal, DOMAIN_CREATE_ACTIONS);
  }
}