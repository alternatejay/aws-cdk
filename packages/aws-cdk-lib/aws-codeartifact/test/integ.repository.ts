#!/usr/bin/env node
import * as iam from '@aws-cdk/aws-iam';
import * as kms from '@aws-cdk/aws-kms';
import * as cdk from '@aws-cdk/core';
import * as codeartifact from '../lib';


const app = new cdk.App();

const stack = new cdk.Stack(app, 'aws-cdk-codeartifact');

const p = new iam.PolicyDocument();
p.addStatements(new iam.PolicyStatement({
  sid: 'DomainPolicyForOrganization',
  effect: iam.Effect.ALLOW,
  actions: [
    'codeartifact:GetDomainPermissionsPolicy',
    'codeartifact:ListRepositoriesInDomain',
    'codeartifact:GetAuthorizationToken',
    'codeartifact:DescribeDomain',
    'codeartifact:CreateRepository',
    'sts:GetServiceBearerToken',
  ],
  principals: [new iam.AnyPrincipal()],
  resources: ['*'],
  conditions: {
    StringEquals: { 'aws:PrincipalOrgID': ['o-12345678910'] },
  },
}));


cdk.Tag.add(stack, 'app:example', 'integ-test-tag');

// Custom KMS key
const key = new kms.Key(stack, 'key');

const domain = new codeartifact.Domain(stack, 'domain', { domainName: 'example-domain', policyDocument: p, domainEncryptionKey: key });
const upstream = new codeartifact.Repository(stack, 'repository', {
  repositoryName: 'repository',
  domain: domain,
  externalConnections: [codeartifact.ExternalConnection.NPM],
});

const subRepo = new codeartifact.Repository(stack, 'sub-repository', {
  repositoryName: 'sub-repository',
  domain: domain,
  upstreams: [upstream],
});

subRepo.allowReadFromRepository(new iam.AccountRootPrincipal());


app.synth();