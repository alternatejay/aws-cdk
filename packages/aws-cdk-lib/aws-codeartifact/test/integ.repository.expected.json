{
  "Resources": {
    "keyFEDD6EC0": {
      "Type": "AWS::KMS::Key",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
                "kms:GenerateDataKey",
                "kms:TagResource",
                "kms:UntagResource"
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":root"
                    ]
                  ]
                }
              },
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "Tags": [
          {
            "Key": "app:example",
            "Value": "integ-test-tag"
          }
        ]
      },
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "domainFBFFA2F6": {
      "Type": "AWS::CodeArtifact::Domain",
      "Properties": {
        "DomainName": "example-domain",
        "EncryptionKey": {
          "Ref": "keyFEDD6EC0"
        },
        "PermissionsPolicyDocument": {
          "Statement": [
            {
              "Action": [
                "codeartifact:GetDomainPermissionsPolicy",
                "codeartifact:ListRepositoriesInDomain",
                "codeartifact:GetAuthorizationToken",
                "codeartifact:DescribeDomain",
                "codeartifact:CreateRepository",
                "sts:GetServiceBearerToken"
              ],
              "Condition": {
                "StringEquals": {
                  "aws:PrincipalOrgID": [
                    "o-12345678910"
                  ]
                }
              },
              "Effect": "Allow",
              "Principal": "*",
              "Resource": "*",
              "Sid": "DomainPolicyForOrganization"
            }
          ],
          "Version": "2012-10-17"
        },
        "Tags": [
          {
            "Key": "app:example",
            "Value": "integ-test-tag"
          }
        ]
      }
    },
    "repository5D02F86B": {
      "Type": "AWS::CodeArtifact::Repository",
      "Properties": {
        "DomainName": {
          "Fn::GetAtt": [
            "domainFBFFA2F6",
            "Name"
          ]
        },
        "RepositoryName": "repository",
        "ExternalConnections": [
          "public:npmjs"
        ],
        "PermissionsPolicyDocument": {
          "Statement": [
            {
              "Action": [
                "codeartifact:DescribePackageVersion",
                "codeartifact:DescribeRepository",
                "codeartifact:GetPackageVersionReadme",
                "codeartifact:GetRepositoryEndpoint",
                "codeartifact:ListPackageVersionAssets",
                "codeartifact:ListPackageVersionDependencies",
                "codeartifact:ListPackageVersions",
                "codeartifact:ListPackages",
                "codeartifact:ReadFromRepository"
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":root"
                    ]
                  ]
                }
              },
              "Resource": "*"
            },
            {
              "Action": [
                "codeartifact:PublishPackageVersion",
                "codeartifact:PutPackageMetadata"
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":root"
                    ]
                  ]
                }
              },
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "Tags": [
          {
            "Key": "app:example",
            "Value": "integ-test-tag"
          }
        ]
      }
    },
    "subrepository65E28DB9": {
      "Type": "AWS::CodeArtifact::Repository",
      "Properties": {
        "DomainName": {
          "Fn::GetAtt": [
            "domainFBFFA2F6",
            "Name"
          ]
        },
        "RepositoryName": "sub-repository",
        "PermissionsPolicyDocument": {
          "Statement": [
            {
              "Action": [
                "codeartifact:DescribePackageVersion",
                "codeartifact:DescribeRepository",
                "codeartifact:GetPackageVersionReadme",
                "codeartifact:GetRepositoryEndpoint",
                "codeartifact:ListPackageVersionAssets",
                "codeartifact:ListPackageVersionDependencies",
                "codeartifact:ListPackageVersions",
                "codeartifact:ListPackages",
                "codeartifact:ReadFromRepository"
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":root"
                    ]
                  ]
                }
              },
              "Resource": "*"
            },
            {
              "Action": [
                "codeartifact:PublishPackageVersion",
                "codeartifact:PutPackageMetadata"
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId"
                      },
                      ":root"
                    ]
                  ]
                }
              },
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "Tags": [
          {
            "Key": "app:example",
            "Value": "integ-test-tag"
          }
        ],
        "Upstreams": [
          "repository"
        ]
      },
      "DependsOn": [
        "repository5D02F86B"
      ]
    }
  }
}